lib/libsys/auxv.3
.\"
.\" Copyright (c) 2019 Ian Lepore <ian@freebsd.org>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd June 24, 2025
.Dt ELF_AUX_INFO 3
.Os
.Sh NAME
.Nm elf_aux_info
.Nd extract data from the elf auxiliary vector of the current process
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In sys/auxv.h
.Ft int
.Fn elf_aux_info "int aux" "void *buf" "int buflen"
.Sh DESCRIPTION
The
.Fn elf_aux_info
function retrieves the auxiliary info vector requested in
.Va aux .
The information is stored into the provided buffer if it will fit.
The following values, defined in
.In sys/elf_common.h
can be requested (corresponding buffer sizes are specified in parenthesis):
.Bl -tag -width AT_OSRELDATE
.It AT_CANARY
The canary value for SSP (arbitrary sized buffer, as many bytes are
returned as it fits into it, rest is zeroed).
.It AT_EXECPATH
The path of executed program
.Dv (MAXPATHLEN).
This may not be present if the process was initialized by
.Xr fexecve 2
and the namecache no longer contains the file's name.
.It AT_HWCAP
CPU / hardware feature flags
.Dv (sizeof(u_long)).
.It AT_HWCAP2
CPU / hardware feature flags
.Dv (sizeof(u_long)).
.It AT_HWCAP3
CPU / hardware feature flags
.Dv (sizeof(u_long)).
.It AT_HWCAP4
CPU / hardware feature flags
.Dv (sizeof(u_long)).
.It AT_NCPUS
Number of CPUs
.Dv (sizeof(int)).
.It AT_OSRELDATE
The
.Dv OSRELDATE
of the kernel or jail the program is running on
.Dv (sizeof(int)).
.It AT_PAGESIZES
Vector of page sizes (arbitrary sized buffer, as many elements of the
.Dv pagesizes
array are returned as it fits).
.It AT_PAGESZ
Page size in bytes
.Dv (sizeof(int)).
.It AT_TIMEKEEP
Pointer to VDSO timehands (for library internal use,
.Dv sizeof(void *)).
.It AT_USRSTACKBASE
Top of the user stack for main thread.
.It AT_USRSTACKLIM
Limit for grow of the user stack for main thread.
.El
.Sh RETURN VALUES
Returns zero on success, or an error number on failure.
.Sh ERRORS
.Bl -tag -width Er
.It Bq Er EINVAL
An unknown item was requested.
.It Bq Er EINVAL
The provided buffer was not the right size for the requested item.
.It Bq Er ENOENT
The requested item is not available.
.El
.Sh HISTORY
The
.Fn elf_aux_info
function appeared in
.Fx 12.0 .
.Sh BUGS
Only a small subset of available auxiliary info vector items are
accessible with this function.
Some items require a "right-sized" buffer while others just require a
"big enough" buffer.


NEWS
May 20, 2025: libarchive 3.8.0 released

Mar 30, 2025: libarchive 3.7.9 released

Mar 20, 2025: libarchive 3.7.8 released

Oct 13, 2024: libarchive 3.7.7 released

Sep 23, 2024: libarchive 3.7.6 released

Sep 13, 2024: libarchive 3.7.5 released

Apr 26, 2024: libarchive 3.7.4 released

Apr 08, 2024: libarchive 3.7.3 released

Sep 12, 2023: libarchive 3.7.2 released

Jul 29, 2023: libarchive 3.7.1 released

Jul 18, 2023: libarchive 3.7.0 released

Jul 14, 2023: bsdunzip port from FreeBSD

Dec 07, 2022: libarchive 3.6.2 released

Apr 08, 2022: libarchive 3.6.1 released

Feb 09, 2022: libarchive 3.6.0 released

Feb 08, 2022: libarchive 3.5.3 released

Aug 22, 2021: libarchive 3.5.2 released

Dec 26, 2020: libarchive 3.5.1 released

Dec 01, 2020: libarchive 3.5.0 released

Oct 14, 2020: Support for system extended attributes

May 20, 2020: libarchive 3.4.3 released

Apr 30, 2020: Support for pzstd compressed files

Apr 16, 2020: Support for RHT.security.selinux tar extended attribute

Feb 11, 2020: libarchive 3.4.2 released

Jan 23, 2020: Important fixes for writing XAR archives

Jan 20, 2020: New tar option: --safe-writes (atomical file extraction)

Jan 03, 2020: Support mbed TLS (PolarSSL) as optional crypto provider

Dec 30, 2019: libarchive 3.4.1 released

Dec 11, 2019: New pax write option "xattrhdr"

Nov 17, 2019: Unicode filename support for reading lha/lzh archives

Jun 11, 2019: libarchive 3.4.0 released

May 18, 2019: Fixes for reading Android APK and JAR archives

Apr 16, 2019: Support for non-recursive list and extract

Apr 14, 2019: New tar option: --exclude-vcs

Mar 27, 2019: Support for file and directory symlinks on Windows

Mar 12, 2019: Important fixes for storing file attributes and flags

Jan 20, 2019: Support for xz, lzma, ppmd8 and bzip2 decompression in ZIP files

Oct 06, 2018: RAR 5.0 reader

Sep 03, 2018: libarchive 3.3.3 released

Jul 19, 2018: Avoid super-linear slowdown on malformed mtree files

Jan 27, 2018: Many fixes for building with Visual Studio

Oct 19, 2017: NO_OVERWRITE doesn't change existing directory attributes

Aug 12, 2017: New support for Zstandard read and write filters

Jul 09, 2017: libarchive 3.3.2 released

Mar 16, 2017: NFSv4 ACL support for Linux (librichacl)

Feb 26, 2017: libarchive 3.3.1 released
    Security & Feature release

Feb 19, 2017: libarchive 3.3.0 released
    Security & Feature release

Jan 29, 2017: Limited NFSv4 ACL support for Mac OS (Darwin)

Jan 10, 2017: POSIX.1e and NFSv4 ACL support for Solaris and derivates

Dec 27, 2016: NFSv4 ACL read and write support for pax
    Deprecated functions: archive_entry_acl_text(), archive_entry_acl_text_w()

Nov, 2016: libarchive is now being tested by the OSS-Fuzz project

Oct 26, 2016: Remove liblzmadec support

Oct 23, 2016: libarchive 3.2.2 released
    Security release

Jun 20, 2016: libarchive 3.2.1 released
    This fixes a handful of security and other critical issues with 3.2.0

May 01, 2016: libarchive 3.2.0 released

Apr 09, 2016: libarchive 3.1.901a released
    Another test release in preparation for 3.2.0

Feb 13, 2016: libarchive 3.1.900a released
    This is a test release in preparation for 3.2.0

Oct 21, 2015: Preliminary port to OSF

Apr 11, 2015: libarchive's issue tracker is now hosted at GitHub.
    https://github.com/libarchive/libarchive/issues

Early 2015:  Many fixes to crash and overflow bugs thanks to Hanno Boeck

Oct 13, 2014: Zip encryption and decryption support

Aug 13, 2014: Add support for lz4 compression.

Jun 10, 2014: Add warc format support

May  3, 2014: Add experimental Zip streaming extension

Apr  6, 2014: Add bsdcat command-line tool

Jan 12, 2014: Add Zip64 support

Dec  1, 2013: Rewrite Zip write logic

Jul  1, 2013: Add ability to detect encrypted entries for many formats
     (This does not add the ability to *decrypt* those entries, however)

Feb 23, 2013: "raw" write support added

Feb 09, 2013: libarchive 3.1.2 released

Jan 28, 2013: libarchive's new website moved to http://www.libarchive.org.

Jan 13, 2013: libarchive 3.1.1 released

Jan 13, 2013: libarchive 3.1.0 released

Dec 07, 2012: Implement functions to manually set the format and filters used.
    
Nov 11, 2012: Add support for __MACOSX directory in Zip archives, which resource
    forks are stored in.

Oct 20, 2012: Add support for writing v7 tar format.

Oct 09, 2012: Add support for grzip compression.

Oct 07, 2012: Introduce b64encode filter.    
Oct 07, 2012: Introduce uuencode filter.

Oct 06, 2012: Add support for lzop.
    
Sep 27, 2012: Implement function used to seek within data blocks.
    (Currently only supported for uncompressed RAR archives).
    
Apr 22, 2012: Add basic archive read and write filter support for lrzip.

Mar 27, 2012: libarchive 3.0.4 released

Feb 05, 2012: libarchive development now hosted at GitHub.
    http://libarchive.github.com/
Feb 05, 2012: libarchive's issue tracker remains at Google Code.
    http://code.google.com/p/libarchive/issues/list
Feb 05, 2012: libarchive's mailing lists remain at Google Groups.

Dec 24, 2011: libarchive 3.0.2 released
Dec 23, 2011: Various fixes merged from FreeBSD
Dec 23, 2011: Symlink support in Zip reader and writer
Dec 23, 2011: Robustness fixes to 7Zip reader

Nov 27, 2011: libarchive 3.0.1b released

Nov 26, 2011: 7Zip reader
Nov 26, 2011: Small fixes to ISO and Zip to improve robustness with corrupted input
Nov 24, 2011: Improve streaming Zip reader's support for uncompressed entries
Nov 20, 2011: New seeking Zip reader supports SFX Zip archives
Nov 20, 2011: Build fixes on Windows

Nov 13, 2011: libarchive 3.0.0a released

Nov 06, 2011: Update shared-library version calculations for libarchive 3.x
Sep 04, 2011: Fix tar -s; follow GNU tar for controlling hardlink/symlink substitutions
Aug 18, 2011: Fix reading ISO images built by NetBSD's mkisofs
Aug 15, 2011: Old archive_read_support_compression_XXX functions are deprecated and
   will disappear in libarchive 4.0.
Jun 26, 2011: RAR reader

lib/libsys/auxv.c
/*-
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * Copyright 2010, 2012 Konstantin Belousov <kib@FreeBSD.ORG>.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include "namespace.h"
#include <elf.h>
#include <errno.h>
#include <link.h>
#include <pthread.h>
#include <stdbool.h>
#include <string.h>
#include <sys/auxv.h>
#include "un-namespace.h"
#include "libc_private.h"
#include <machine/atomic.h>

extern int _DYNAMIC;
#pragma weak _DYNAMIC

void *__elf_aux_vector;

#ifndef PIC
static pthread_once_t aux_vector_once = PTHREAD_ONCE_INIT;

static void
init_aux_vector_once(void)
{
	Elf_Addr *sp;

	sp = (Elf_Addr *)environ;
	while (*sp++ != 0)
		;
	__elf_aux_vector = (Elf_Auxinfo *)sp;
}

void
__init_elf_aux_vector(void)
{

	if (&_DYNAMIC != NULL)
		return;
	_once(&aux_vector_once, init_aux_vector_once);
}
#endif

static int aux_once;
static int pagesize, osreldate, canary_len, ncpus, pagesizes_len, bsdflags;
static int hwcap_present, hwcap2_present;
static char *canary, *pagesizes, *execpath;
static void *ps_strings, *timekeep;
static u_long hwcap, hwcap2;
static void *fxrng_seed_version;
static u_long usrstackbase, usrstacklim;

#ifdef __powerpc__
static int powerpc_new_auxv_format = 0;
static void _init_aux_powerpc_fixup(void);
int _powerpc_elf_aux_info(int, void *, int);
#endif

/*
 * This function might be called and actual body executed more than
 * once in multithreading environment.  Due to this, it is and must
 * continue to be idempotent.  All stores are atomic (no store
 * tearing), because we only assign to int/long/ptr.
 */
static void
init_aux(void)
{
	Elf_Auxinfo *aux;

	if (atomic_load_acq_int(&aux_once))
		return;
	for (aux = __elf_aux_vector; aux->a_type != AT_NULL; aux++) {
		switch (aux->a_type) {
		case AT_BSDFLAGS:
			bsdflags = aux->a_un.a_val;
			break;

		case AT_CANARY:
			canary = (char *)(aux->a_un.a_ptr);
			break;

		case AT_CANARYLEN:
			canary_len = aux->a_un.a_val;
			break;

		case AT_EXECPATH:
			execpath = (char *)(aux->a_un.a_ptr);
			break;

		case AT_HWCAP:
			hwcap_present = 1;
			hwcap = (u_long)(aux->a_un.a_val);
			break;

		case AT_HWCAP2:
			hwcap2_present = 1;
			hwcap2 = (u_long)(aux->a_un.a_val);
			break;

		case AT_PAGESIZES:
			pagesizes = (char *)(aux->a_un.a_ptr);
			break;

		case AT_PAGESIZESLEN:
			pagesizes_len = aux->a_un.a_val;
			break;

		case AT_PAGESZ:
			pagesize = aux->a_un.a_val;
			break;

		case AT_OSRELDATE:
			osreldate = aux->a_un.a_val;
			break;

		case AT_NCPUS:
			ncpus = aux->a_un.a_val;
			break;

		case AT_TIMEKEEP:
			timekeep = aux->a_un.a_ptr;
			break;

		case AT_PS_STRINGS:
			ps_strings = aux->a_un.a_ptr;
			break;

		case AT_FXRNG:
			fxrng_seed_version = aux->a_un.a_ptr;
			break;

		case AT_USRSTACKBASE:
			usrstackbase = aux->a_un.a_val;
			break;

		case AT_USRSTACKLIM:
			usrstacklim = aux->a_un.a_val;
			break;
#ifdef __powerpc__
		/*
		 * Since AT_STACKPROT is always set, and the common
		 * value 23 is mutually exclusive with the legacy powerpc
		 * value 21, the existence of AT_STACKPROT proves we are
		 * on the common format.
		 */
		case AT_STACKPROT:	/* 23 */
			powerpc_new_auxv_format = 1;
			break;
#endif
		}
	}
#ifdef __powerpc__
	if (!powerpc_new_auxv_format)
		_init_aux_powerpc_fixup();
#endif

	atomic_store_rel_int(&aux_once, 1);
}

#ifdef __powerpc__
static void
_init_aux_powerpc_fixup(void)
{
	Elf_Auxinfo *aux;

	/*
	 * Before 1300070, PowerPC platforms had nonstandard numbering for
	 * the aux vector. When running old binaries, the kernel will pass
	 * the vector using the old numbering. Reload affected variables.
	 */
	for (aux = __elf_aux_vector; aux->a_type != AT_NULL; aux++) {
		switch (aux->a_type) {
		case AT_OLD_CANARY:
			canary = (char *)(aux->a_un.a_ptr);
			break;
		case AT_OLD_CANARYLEN:
			canary_len = aux->a_un.a_val;
			break;
		case AT_OLD_EXECPATH:
			execpath = (char *)(aux->a_un.a_ptr);
			break;
		case AT_OLD_PAGESIZES:
			pagesizes = (char *)(aux->a_un.a_ptr);
			break;
		case AT_OLD_PAGESIZESLEN:
			pagesizes_len = aux->a_un.a_val;
			break;
		case AT_OLD_OSRELDATE:
			osreldate = aux->a_un.a_val;
			break;
		case AT_OLD_NCPUS:
			ncpus = aux->a_un.a_val;
			break;
		}
	}
}

int
_powerpc_elf_aux_info(int aux, void *buf, int buflen)
{

	/*
	 * If we are in the old auxv format, we need to translate the aux
	 * parameter of elf_aux_info() calls into the common auxv format.
	 * Internal libc calls always use the common format, and they
	 * directly call _elf_aux_info instead of using the weak symbol.
	 */
	if (!powerpc_new_auxv_format) {
		switch (aux) {
		case AT_OLD_EXECPATH:
			aux = AT_EXECPATH;
			break;
		case AT_OLD_CANARY:
			aux = AT_CANARY;
			break;
		case AT_OLD_CANARYLEN:
			aux = AT_CANARYLEN;
			break;
		case AT_OLD_OSRELDATE:
			aux = AT_OSRELDATE;
			break;
		case AT_OLD_NCPUS:
			aux = AT_NCPUS;
			break;
		case AT_OLD_PAGESIZES:
			aux = AT_PAGESIZES;
			break;
		case AT_OLD_PAGESIZESLEN:
			aux = AT_PAGESIZESLEN;
			break;
		case AT_OLD_STACKPROT:
			aux = AT_STACKPROT;
			break;
		}
	}
	return _elf_aux_info(aux, buf, buflen);
}
__weak_reference(_powerpc_elf_aux_info, elf_aux_info);
#else
__weak_reference(_elf_aux_info, elf_aux_info);
#endif

int
_elf_aux_info(int aux, void *buf, int buflen)
{
	int res;

#ifndef PIC
	__init_elf_aux_vector();
#endif
	if (__elf_aux_vector == NULL)
		return (ENOSYS);
	init_aux();	/* idempotent */

	if (buflen < 0)
		return (EINVAL);

	switch (aux) {
	case AT_CANARY:
		if (canary != NULL && canary_len >= buflen) {
			memcpy(buf, canary, buflen);
			memset(canary, 0, canary_len);
			canary = NULL;
			res = 0;
		} else
			res = ENOENT;
		break;
	case AT_EXECPATH:
		if (execpath == NULL)
			res = ENOENT;
		else if (buf == NULL)
			res = EINVAL;
		else {
			if (strlcpy(buf, execpath, buflen) >=
			    (unsigned int)buflen)
				res = EINVAL;
			else
				res = 0;
		}
		break;
	case AT_HWCAP:
		if (hwcap_present && buflen == sizeof(u_long)) {
			*(u_long *)buf = hwcap;
			res = 0;
		} else
			res = ENOENT;
		break;
	case AT_HWCAP2:
		if (hwcap2_present && buflen == sizeof(u_long)) {
			*(u_long *)buf = hwcap2;
			res = 0;
		} else
			res = ENOENT;
		break;
	case AT_PAGESIZES:
		if (pagesizes != NULL && pagesizes_len >= buflen) {
			memcpy(buf, pagesizes, buflen);
			res = 0;
		} else
			res = ENOENT;
		break;
	case AT_PAGESZ:
		if (buflen == sizeof(int)) {
			if (pagesize != 0) {
				*(int *)buf = pagesize;
				res = 0;
			} else
				res = ENOENT;
		} else
			res = EINVAL;
		break;
	case AT_OSRELDATE:
		if (buflen == sizeof(int)) {
			if (osreldate != 0) {
				*(int *)buf = osreldate;
				res = 0;
			} else
				res = ENOENT;
		} else
			res = EINVAL;
		break;
	case AT_NCPUS:
		if (buflen == sizeof(int)) {
			if (ncpus != 0) {
				*(int *)buf = ncpus;
				res = 0;
			} else
				res = ENOENT;
		} else
			res = EINVAL;
		break;
	case AT_TIMEKEEP:
		if (buflen == sizeof(void *)) {
			if (timekeep != NULL) {
				*(void **)buf = timekeep;
				res = 0;
			} else
				res = ENOENT;
		} else
			res = EINVAL;
		break;
	case AT_BSDFLAGS:
		if (buflen == sizeof(int)) {
			*(int *)buf = bsdflags;
			res = 0;
		} else
			res = EINVAL;
		break;
	case AT_PS_STRINGS:
		if (buflen == sizeof(void *)) {
			if (ps_strings != NULL) {
				*(void **)buf = ps_strings;
				res = 0;
			} else
				res = ENOENT;
		} else
			res = EINVAL;
		break;
	case AT_FXRNG:
		if (buflen == sizeof(void *)) {
			if (fxrng_seed_version != NULL) {
				*(void **)buf = fxrng_seed_version;
				res = 0;
			} else
				res = ENOENT;
		} else
			res = EINVAL;
		break;
	case AT_USRSTACKBASE:
		if (buflen == sizeof(u_long)) {
			if (usrstackbase != 0) {
				*(u_long *)buf = usrstackbase;
				res = 0;
			} else
				res = ENOENT;
		} else
			res = EINVAL;
		break;
	case AT_USRSTACKLIM:
		if (buflen == sizeof(u_long)) {
			if (usrstacklim != 0) {
				*(u_long *)buf = usrstacklim;
				res = 0;
			} else
				res = ENOENT;
		} else
			res = EINVAL;
		break;
	default:
		res = ENOENT;
		break;
	}
	return (res);
}
